packages/prover/src/web3_provider_inspector.ts

```diff
@@ -0,0 +1,89 @@
+import {Logger} from "@lodestar/logger";
+import {AnyWeb3Provider, Web3ProviderType} from "./interfaces.js";
+
+import web3jsProviderType from "./provider_types/web3_js_provider_type.js";
+import ethersProviderType from "./provider_types/ethers_provider_type.js";
+import eip1193ProviderType from "./provider_types/eip1193_provider_type.js";
+import legacyProviderType from "./provider_types/legacy_provider_type.js";
+
+export class Web3ProviderInspector {
+  protected providerTypes: Web3ProviderType<AnyWeb3Provider>[] = [];
+  logger: Logger;
+
+  protected constructor(opts: {logger: Logger}) {
+    this.logger = opts.logger;
+  }
+
+  static initWithDefault(opts: {logger: Logger}): Web3ProviderInspector {
+    const inspector = new Web3ProviderInspector(opts);
+    inspector.register(web3jsProviderType, {index: 0});
+    inspector.register(ethersProviderType, {index: 1});
+    inspector.register(eip1193ProviderType, {index: 2});
+    inspector.register(legacyProviderType, {index: 3});
+
+    return inspector;
+  }
+
+  getProviderTypes(): Web3ProviderType<AnyWeb3Provider>[] {
+    // Destruct so user can not mutate the output
+    return [...this.providerTypes];
+  }
+
+  register(providerType: Web3ProviderType<AnyWeb3Provider>, opts?: {index?: number}): void {
```

Comment:
Is there a use-case for user provided indexes? Or maybe register could accept an array directly?

Reply:
A provdier e.g. have `request` method which could be identified as `EIP1193` provider, but the implementation of that provdier could be different. 

So user would like to add their custom provider logic at earlier index. The first one matches will be returned.

Reply:
Makes sense. An idea to simplify would be to always prepend then, and allow user to register an ordered collection of `providers`

Reply:
From user facing function `createVerifiedExecutionProvider` we do exactly the same, user don't interact with the `Web3ProviderInspector` directly, it's an internal API.